src/charm.py

```diff
@@ -572,13 +686,22 @@ def has_config_server(self) -> bool:
     @property
     def expose_external(self) -> Optional[str]:
         """Returns mode of exposure for external connections."""
-        if (
-            self.app_peer_data.get("expose-external", Config.ExternalConnections.NONE)
-            == Config.ExternalConnections.NONE
-        ):
+        # don't let an incorrect configuration
+        if not self.is_user_external_config_valid():
+            if (
+                self.app_peer_data.get(
+                    "expose-external", Config.ExternalConnections.NONE
+                )
+                == Config.ExternalConnections.NONE
+            ):
+                return None
+
+            return self.app_peer_data["expose-external"]
```

Comment:
**question:** What is this for? Just curious why you're looking at app-data, what case are we covering here?

Reply:
This covers the case that the user has incorrectly set the value of `expose-external`, i.e. they did `juju config mongos-k8s expose-external=this-is-totally-not-valid`, when they do this we don't want their incorrect setting to break operations in the charm. When this happens we put the charm into blocked state and then use the last valid value (which should be saved in the app databag))

Reply:
Gotcha, I guess in our charms we just raise, but makes sense.